File: packages/sdk-go/runback/events.go
package runback

// Wire format for /api/ingest — Phase 1 subset of packages/schema/src/events.ts.
// Field names are snake_case to match the wire schema exactly; Go's JSON tags
// carry the translation instead of a separate marshaling layer.
//
// Phase 1 covers event capture, ingest, client-side redaction (see redact.go,
// ported from packages/redact/src/patterns.ts, on by default at the standard
// tier), and cassette hash-chaining (see cassette.go, ported from
// packages/replay/src/cassette.ts) — a Go-recorded run gets a real
// cassette_digest, cross-checked byte-for-byte against the TS implementation
// in cassette_test.go. What's NOT ported: packages/replay/src/replay.ts's
// re-execution engine, so a Go-recorded run's digest proves tamper-evidence
// but can't be stepped through/bisected the way a TS- or Python-recorded run
// can. That's the real remaining gap for a future phase.

const SchemaVersion = 1

// Actor identifies who or what triggered an event — a specific end-user, a
// service/API-key identity, or an automated schedule.
type Actor struct {
	Type  string `json:"type"` // "user" | "api_key" | "system"
	ID    string `json:"id"`
	Label string `json:"label,omitempty"`
}

// TraceError is the wire shape for a captured error.
type TraceError struct {
	Name    string `json:"name"`
	Message string `json:"message"`
	Stack   string `json:"stack,omitempty"`
}

// Model identifies the provider and model_id of an LLM call.
type Model struct {
	Provider string `json:"provider"`
	ModelID  string `json:"model_id"`
}

// ModelMessage is one message in the request sent to the model.
type ModelMessage struct {
	Role    string `json:"role"` // "system" | "user" | "assistant" | "tool"
	Content any    `json:"content"`
}

// ToolDefinition is a tool exposed to the model in an LLM request.
type ToolDefinition struct {
	Name        string `json:"name"`
	Description string `json:"description,omitempty"`
	Parameters  any    `json:"parameters,omitempty"`
}

// ModelParams are the sampling parameters sent with an LLM request.
type ModelParams struct {
	Temperature     *float64 `json:"temperature,omitempty"`
	MaxOutputTokens *float64 `json:"max_output_tokens,omitempty"`
	TopP            *float64 `json:"top_p,omitempty"`
}

// LlmRequest is the exact context window sent to the model.
type LlmRequest struct {
	System   *string          `json:"system"`
	Messages []ModelMessage   `json:"messages"`
	Tools    []ToolDefinition `json:"tools"`
	Params   ModelParams      `json:"params"`
}

// ToolCall is a tool the model asked to invoke, as returned in an LLM response.
type ToolCall struct {
	ToolCallID string `json:"tool_call_id"`
	ToolName   string `json:"tool_name"`
	Input      any    `json:"input,omitempty"`
}

// LlmResponse is what the model returned.
type LlmResponse struct {
	Text         *string    `json:"text"`
	Reasoning    *string    `json:"reasoning"`
	FinishReason *string    `json:"finish_reason"`
	ToolCalls    []ToolCall `json:"tool_calls"`
}

// TokenUsage is the token accounting for one LLM call.
type TokenUsage struct {
	InputTokens  int `json:"input_tokens"`
	OutputTokens int `json:"output_tokens"`
	TotalTokens  int `json:"total_tokens"`
}

// baseEvent carries the fields common to every event type. Each concrete
// event type embeds it; encoding/json promotes its fields to the top level
// of the marshaled object, matching the flat wire shape.
type baseEvent struct {
	SchemaVersion int     `json:"schema_version"`
	RunID         string  `json:"run_id"`
	SpanID        string  `json:"span_id"`
	ParentSpanID  *string `json:"parent_span_id"`
	Seq           int     `json:"seq"`
	TsStart       string  `json:"ts_start"`
	TsEnd         *string `json:"ts_end"`
	Actor         *Actor  `json:"actor,omitempty"`
}

// LlmEvent is a single call to a language model.
type LlmEvent struct {
	baseEvent
	Type      string      `json:"type"` // always "llm"
	Model     Model       `json:"model"`
	Request   LlmRequest  `json:"request"`
	Response  LlmResponse `json:"response"`
	Usage     *TokenUsage `json:"usage"`
	LatencyMs *int        `json:"latency_ms"`
	Error     *TraceError `json:"error"`
}

// ToolEvent is a single tool execution.
type ToolEvent struct {
	baseEvent
	Type       string      `json:"type"` // always "tool"
	ToolName   string      `json:"tool_name"`
	ToolCallID string      `json:"tool_call_id"`
	Input      any         `json:"input,omitempty"`
	Output     any         `json:"output,omitempty"`
	LatencyMs  *int        `json:"latency_ms"`
	Error      *TraceError `json:"error"`
}

// ReasoningEvent is a captured intermediate thought, independent of any
// specific model call — e.g. a routing decision in a graph-based agent.
type ReasoningEvent struct {
	baseEvent
	Type  string  `json:"type"` // always "reasoning"
	Text  string  `json:"text"`
	Label *string `json:"label"`
}

// RunEvent is the start or end envelope for one agent invocation.
type RunEvent struct {
	baseEvent
	Type     string         `json:"type"`  // always "run"
	Phase    string         `json:"phase"` // "start" | "end"
	Name     string         `json:"name"`
	Input    any            `json:"input,omitempty"`
	Output   any            `json:"output,omitempty"`
	Status   string         `json:"status"`
	Error    *TraceError    `json:"error"`
	Metadata map[string]any `json:"metadata"`
}

File: packages/schema/src/events.ts
/**
 * Runback trace-event wire format — v1.
 *
 * Design principles (these are load-bearing; read before changing):
 *  1. FLAT event stream, not a nested tree. Events carry `parent_span_id` and the
 *     tree is reconstructed on the client. A flat stream ingests incrementally,
 *     survives partial failures, and keeps an OpenTelemetry export path open.
 *  2. DISCRIMINATED UNION on `type` — one exhaustively-switchable renderer type.
 *  3. EVERYTHING ADDRESSABLE — every event has a stable (`run_id`, `span_id`) so a
 *     future "replay from step N" / "breakpoint on span X" has a durable anchor.
 *
 * Append-only contract: never repurpose a field, only add OPTIONAL ones, and bump
 * `schema_version` for breaking changes.
 */

export type SchemaVersion = 1;
export const SCHEMA_VERSION: SchemaVersion = 1;

export interface TraceError {
  name: string;
  message: string;
  stack?: string;
}

export interface ModelMessage {
  role: "system" | "user" | "assistant" | "tool";
  /** string OR a content-parts array (multimodal / tool-result parts). */
  content: unknown;
}

export interface ToolDefinition {
  name: string;
  description?: string;
  /** JSON Schema for the tool's parameters, as exposed to the model. */
  parameters: unknown;
}

export interface TokenUsage {
  input_tokens: number;
  output_tokens: number;
  total_tokens: number;
}

/**
 * Salience projection (H2) — how a tool/model call's input is normalized BEFORE
 * it is content-addressed, so the same logical call keys identically across runs
 * even when volatile fields differ (request ids, timestamps, nonces). Carried on
 * the event so the server reproduces the exact projection and recomputes the key
 * from the RAW input — altering a salient field still breaks the chain; declared-
 * volatile fields are simply excluded from identity. Dot-paths, array-aware.
 * Absent ⇒ the key is computed over the raw input, exactly as before.
 */
export interface KeyProjection {
  /** Top-level allowlist: keep only these top-level fields when keying (subtrees intact). */
  keep?: string[];
  /** Dot-path denylist (array-aware): drop these volatile fields anywhere before keying. */
  drop?: string[];
}

export interface BaseEvent {
  schema_version: SchemaVersion;
  /** One agent invocation. */
  run_id: string;
  /** Unique id for THIS event (ULID — lexicographically sortable). */
  span_id: string;
  /** Tree reconstruction. `null` = root. */
  parent_span_id: string | null;
  /** Monotonic per-run ordering. Survives clock skew; use this to sort, not timestamps. */
  seq: number;
  /** ISO 8601. */
  ts_start: string;
  /** ISO 8601, or `null` while in-flight (forward-compat for live streaming). */
  ts_end: string | null;
  type: TraceEvent["type"];
  /**
   * Who or what triggered THIS event — a specific end-user, a service/API-key
   * identity, or an automated schedule. Optional and self-reported by the
   * caller (like `tags`); set once per Collector via CollectorOptions.actor,
   * or overridden per call. Answers "who caused this decision," not just
   * "what happened."
   */
  actor?: { type: "user" | "api_key" | "system"; id: string; label?: string };
}

/** A single call to a language model. The `request` field is the whole point of Runback. */
export interface LlmEvent extends BaseEvent {
  type: "llm";
  model: { provider: string; model_id: string };
  /**
   * THE money field — the exact context window the model saw (request side).
   * Captured verbatim at the `wrapLanguageModel` middleware boundary, NOT
   * reconstructed from response callbacks.
   */
  request: {
    system: string | null;
    messages: ModelMessage[];
    tools: ToolDefinition[];
    params: {
      temperature?: number;
      max_output_tokens?: number;
      top_p?: number;
    };
  };
  response: {
    text: string | null;
    reasoning: string | null;
    finish_reason: string | null;
    tool_calls: { tool_call_id: string; tool_name: string; input: unknown }[];
  };
  usage: TokenUsage | null;
  latency_ms: number | null;
  error: TraceError | null;
  /** Optional salience projection applied when content-addressing this request. */
  key_projection?: KeyProjection;
  /**
   * Free-form, self-reported context — e.g. which graph node/step this call
   * happened inside, for a framework (LangGraph and similar) whose execution
   * unit is bigger than one LLM call. Mirrors RunEvent.metadata; absent for
   * every existing integration, so this is purely additive.
   */
  metadata?: Record<string, unknown>;
}

/** A tool/function invocation requested by an LLM step. */
export interface ToolEvent extends BaseEvent {
  type: "tool";
  tool_name: string;
  /** Links back to the LlmEvent.response.tool_calls entry that requested it. */
  tool_call_id: string;
  input: unknown;
  /** Result, or `null` if it errored / is in-flight. */
  output: unknown | null;
  latency_ms: number | null;
  /** The tool threw — the #1 thing developers debug. */
  error: TraceError | null;
  /** Optional salience projection applied when content-addressing this input. */
  key_projection?: KeyProjection;
  /**
   * Set when a runtime policy guardrail BLOCKED this action before it ran — the
   * tool never executed (`output` is null). A first-class, hash-chained, re-runnable
   * record that the control fired: "we stopped it, and here's the proof".
   */
  policy_block?: { rule: string; detail: string };
  /**
   * Set whenever a runtime policy check actually ran against this call before
   * it executed — evidence the enforcement engine evaluated it, not just that
   * it was blocked. `passed: false` calls also carry `policy_block` above.
   */
  policy_evaluated?: { passed: boolean };
  /** Free-form, self-reported context — see LlmEvent.metadata. */
  metadata?: Record<string, unknown>;
}

/** A reasoning / log marker — manually emitted or extracted from a step. */
export interface ReasoningEvent extends BaseEvent {
  type: "reasoning";
  text: string;
  label: string | null;
  /**
   * Graph-node identity, when the source framework has one and reports it
   * (e.g. OpenInference's generic graph.node.* span attributes) — a plain
   * CHAIN/AGENT-kind span collapses to `text` alone otherwise. Absent for
   * every default integration today (including the standard LangChain/
   * LangGraph OTel instrumentor, which does not emit these attributes);
   * present when a caller's own instrumentation sets them.
   */
  graph_node?: { name: string; step?: number; graph_name?: string };
  /** Predecessor node name(s) — the routing decision, when knowable. */
  routed_from?: string[];
}

/** Run lifecycle envelope. One `start` and (usually) one `end` per run. */
export interface RunEvent extends BaseEvent {
  type: "run";
  phase: "start" | "end";
  name: string;
  input: unknown | null;
  output: unknown | null;
  status: "running" | "success" | "error" | null;
  error: TraceError | null;
  metadata: Record<string, unknown>;
}

/**
 * A nondeterminism boundary the agent crossed — a clock read, a random draw, a
 * generated UUID, or a network fetch — captured in-process by the SDK so a run can
 * be reproduced byte-exact, not just its LLM/tool oracle. The substrate of
 * deterministic replay; see docs/DEEP_REPLAY_SPEC.md.
 */
export interface EnvEvent extends BaseEvent {
  type: "env";
  kind: "now" | "date" | "random" | "uuid" | "fetch";
  /** Content address of the read (for fetch: method+url; for primitives: the kind). */
  key: string;
  /** The value handed back to the computation (the recorded oracle value). */
  output: unknown;
}

export type TraceEvent = LlmEvent | ToolEvent | ReasoningEvent | RunEvent | EnvEvent;

/** Body shape POSTed to the ingest endpoint. */
export interface IngestPayload {
  events: TraceEvent[];
}

/** Exhaustiveness helper for switch statements over `TraceEvent["type"]`. */
export function assertNever(x: never): never {
  throw new Error(`Unexpected trace event variant: ${JSON.stringify(x)}`);
}

/**
 * What a model decided at one step — the shape a replay compares against its
 * recording to judge "same behaviour" or "diverged".
 *
 * Lives in @runback/schema, not in the replay engine that produces it, because
 * both tiers need it: the Community divergence scorer consumes it, while the
 * commercially-licensed re-execution engine constructs it. Defining it beside
 * the engine forced a Community → Enterprise import that broke a
 * Community-only build (caught by scripts/verify-community-build.sh). This is
 * the cross-package type vocabulary, so it belongs here.
 */
export interface ModelDecision {
  tool_calls: { tool_name: string; input: unknown }[];
  finish_reason: string | null;
  text: string | null;
}
